### ISSUE TITLE
Silent Triton multi-output fusion miscompile on consumer Blackwell (sm_120)

### ISSUE THREAD

**Issue by xuanlinli17** [jax-ml/jax#39486]
## Summary

On a **consumer Blackwell GPU (compute capability 12.0, `sm_120`)**  — XLA routes a specific pattern to a `kind = __triton`
fusion that produces **silently wrong results**: no crash, no NaN, no
non-determinism. The output is off by a **large** margin (relative error up to
**2.8e+06** against a float64 reference), fully bit-deterministic across runs.
XLA's own `--xla_gpu_verify_triton_fusion_numerics=true` verifier independently
confirms the fusion is wrong (`triton_fusion_numerics_verifier.cc:270: Output
buffer does not match reference buffer`).

The bug reproduces on jax/jaxlib **0.10.2** and **0.11.0**. jax 0.11.0's more
aggressive fusion makes the pattern reachable far more often, which is when it
starts being hit in practice.

## Environment

| | |
|---|---|
| GPU | NVIDIA consumer Blackwell, compute capability **12.0** (`sm_120`/`sm_120a`) |
| Driver | 595.71.05 |
| jax / jaxlib | reproduced on **0.10.2** and **0.11.0** |
| ptxas | 12.9.86 |

Not reproduced on sm_89 cards with jax 0.5.0 or jax 0.8.3 — there XLA does not form the
offending fusion (the 512×2048 concatenate stays in a `kInput` fusion; the only
`triton_softmax` fusion is standalone and correct).

## Reproducer

Self-contained, ~80 lines, depends only on `jax` + `numpy`:

```python
"""JAX-level repro for the jax 0.11.0 / sm_120 Triton fusion miscompile.

Mirrors the fusion XLA dumped as
  module_XXXX.jit_wrapper.triton_fusion_numerics_verifier_failed_*.hlo
"""
import numpy as np
import jax
import jax.numpy as jnp

print("jax", jax.__version__, "backend", jax.default_backend(), flush=True)

R = 512      # rows
C = 512      # cols
K = 4        # leading dim of the 3-D operand that gets reduced
EPS = 1e-5


def _ln(x, gamma, beta):
    """Layernorm over the last axis."""
    mean = x.mean(axis=-1, keepdims=True)
    xc = x - mean
    var = (xc * xc).mean(axis=-1, keepdims=True)
    return xc / jnp.sqrt(var + EPS) * gamma + beta


def block(p, x3):
    """Two layernorm branches concatenated -> [R, 4*C]."""
    # branch 1: slice a bigger weight, bias, normalize, concat
    w = jax.lax.slice(p["big"], (0, 0), (C, C))          # [512,512] from [1024,512]
    h1 = w + p["b1"]
    h1 = h1 / jnp.sqrt(p["v1"] + EPS)
    left = jnp.concatenate([p["other"], h1], axis=1)      # [512,1024]
    left = left / jnp.sqrt(p["v2"] + EPS)[:, None]

    # branch 2: reduce a 3-D operand, then a real layernorm
    h2 = x3.sum(axis=0) + p["b2"]                         # [512,512]
    right = _ln(h2, p["gamma"], p["beta"])                # [512,512]

    # root concat: 1024 + 512 + 512 = 2048
    return jnp.concatenate([left, right, p["extra"]], axis=1)


def make_params(rng):
    f = lambda *s: jnp.asarray(rng.normal(0, 0.5, size=s), dtype=jnp.float32)
    return {
        "big": f(2 * C, C),
        "other": f(R, C),
        "extra": f(R, C),
        "b1": f(C),
        "b2": f(C),
        "gamma": f(C),
        "beta": f(C),
        "v1": jnp.abs(f(C)) + 0.1,
        "v2": jnp.abs(f(R)) + 0.1,
    }


def ref_block(p, x3):
    """float64 numpy reference."""
    g = lambda k: np.asarray(p[k], dtype=np.float64)
    w = g("big")[:C, :C]
    h1 = (w + g("b1")) / np.sqrt(g("v1") + EPS)
    left = np.concatenate([g("other"), h1], axis=1)
    left = left / np.sqrt(g("v2") + EPS)[:, None]
    h2 = np.asarray(x3, np.float64).sum(axis=0) + g("b2")
    mean = h2.mean(axis=-1, keepdims=True)
    xc = h2 - mean
    var = (xc * xc).mean(axis=-1, keepdims=True)
    right = xc / np.sqrt(var + EPS) * g("gamma") + g("beta")
    return np.concatenate([left, right, g("extra")], axis=1)


rng = np.random.default_rng(0)
p = make_params(rng)
x3 = jnp.asarray(rng.normal(0, 1.0, size=(K, R, C)), dtype=jnp.float32)
ref = ref_block(p, x3)

out = np.asarray(jax.jit(block)(p, x3))
mx_abs = float(np.nanmax(np.abs(out - ref)))
rel = np.abs(out - ref) / np.maximum(np.abs(ref), 1e-30)
mx_rel = float(np.nanmax(rel))
nan = int(np.isnan(out).sum())
ok = nan == 0 and np.all((np.abs(out - ref) < 1e-4) | (rel < 1e-4))
print(f"nan={nan} max_abs={mx_abs:.3e} max_rel={mx_rel:.3e} -> {'OK' if ok else 'BAD'}")
```

### Expected (correct) output

On non-sm_120 hardware, or with the workaround below, the GPU matches the
float64 reference to float32 precision:

```
nan=0 max_abs=3.5e-06 max_rel=2.2e-02 -> OK
```

(`max_rel` ~2e-2 is a metric artifact on the mean-subtracted ~0 layernorm
outputs — an exactly-f32-correct element near zero has unbounded relative
error. `max_abs` ~3.5e-6 and the per-element OR gate confirm correctness.)

### Actual (buggy) output on sm_120 / jax 0.11.0

```
jax 0.11.0 backend gpu
device: <consumer Blackwell, compute capability 12.0>
nan=0 max_abs=~3e+00 max_rel=2.830e+06 -> BAD
```

The output is wrong by **O(1) in absolute terms** (not a precision issue) and
silently so.

## XLA's own numerics verifier confirms it (and dumps the failing fusion)

Run the reproducer with XLA's Triton-fusion numerics verifier enabled, and point
`--xla_dump_to` at a directory so the failing fusion's HLO is captured:

```bash
mkdir -p /tmp/xla_dump
XLA_FLAGS="--xla_dump_to=/tmp/xla_dump \
           --xla_gpu_verify_triton_fusion_numerics=true" \
  python repro.py
```

On sm_120 this prints:

```
E triton_fusion_numerics_verifier.cc:270]
  Triton numerics verification failed with: Output buffer does not match reference buffer.
```

and writes the offending fusion's HLO into the dump directory as a file whose
name matches `*triton_fusion_numerics_verifier_failed_*.hlo` (full pattern:
`module_XXXX.jit_wrapper.triton_fusion_numerics_verifier_failed_*.hlo`). If
`--xla_dump_to` is not set, XLA writes dumps to the current working directory
instead.

That dumped fusion can be replayed in isolation to re-confirm the failure
without re-running the Python graph:

```bash
run_hlo_module --platform=CUDA \
  --xla_gpu_verify_triton_fusion_numerics=true \
  /tmp/xla_dump/*triton_fusion_numerics_verifier_failed_*.hlo
```

## The failing fusion

The fusion XLA dumps (per the step above) has the following backend config:

```
kind = __triton
block_level_fusion_config = {
    num_warps: 8,
    output_tiles: [{ sizes: ["8", "512"] }],
    num_ctas: 1, num_stages: 1
}
```

Shape of the computation:

```
LN1: slice(w[1024,512])[0:512] + b1   -> / sqrt(v1 + 1e-5)
     concat(other[512,512], that)      -> f32[512,1024]
     -> / sqrt(v2 + 1e-5)

LN2: reduce(x[4,512,512], dim=0) + b2 -> x - mean -> square -> sum
     -> / sqrt(var + 1e-5) * gamma + beta   -> f32[512,512]

ROOT concatenate(LN1, LN2, extra)     -> f32[512,2048]
```

### Ingredients that appear necessary to trigger it

- **two independent** reduction chains inside a single `__triton` fusion
- a `concatenate` feeding another `concatenate` (nested concatenates)
- output row width (2048) **much larger than the tile width** (512)

Removing any one of these makes the fusion compile correctly.

## Workaround

Disabling Triton fusion avoids it:

```bash
export XLA_FLAGS="--xla_gpu_enable_triton_gemm=false \
  --xla_gpu_unsupported_enable_triton_multi_output_fusion=false \
  --xla_gpu_experimental_enable_triton_heroless_priority_fusion=false"
```

## Related

- [**#39336**](https://github.com/jax-ml/jax/issues/39336) — *"jit silently miscompiles `alpha * (A @ B + bias)` on sm_120"* (open,
  same sm_120 / jax-0.10.x family: silent, deterministic, large-magnitude Triton
  miscompile). Different root cause (GEMM-epilogue scale vs. multi-output
  reduction-fusion here), but the similarity suggests a broader sm_120 Triton
  correctness gap worth investigating together.

### System info (python version, jaxlib version, accelerator, etc.)

jax:    0.11.0
jaxlib: 0.11.0
numpy:  2.5.1
ptxas  12.9.86
nvidia driver 595.71.05
python: 3.12.13 | packaged by Anaconda, Inc. | (main, Jul  9 2026, 14:38:16) [GCC 14.3.0]
process_count: 1
platform: uname_result(system='Linux', release='6.8.0-71-generic', version='#71-Ubuntu SMP PREEMPT_DYNAMIC Tue Jul 22 16:52:38 UTC 2025', machine='x86_64')

**Comment by olegshyshkov** [comment]
Thank you for reporting! We can reproduce the bug internally and will submit a fix shortly.

**Comment by akuegel** [comment]
Should be fixed with https://github.com/openxla/xla/commit/5d74754e9e638f5fa2e5f67283cd5eb30f7b6d87

**Comment by akuegel** [comment]
@xuanlinli17 Should we keep this issue open until the next JAX release with the fix, or can it be closed already?

**Comment by xuanlinli17** [comment]
Probably keep it open until next jax release? But it's up to the jax developers.
